File: solutions/3517-shortest-distance-after-road-addition-queries-i/shortest-distance-after-road-addition-queries-i.py
class Solution:
    def dfs(self, tree, c, dp):
        d = dp[c] + 1
        for x in tree[c]:
            if dp[x] <= d:
                continue
            dp[x] = d
            self.dfs(tree, x, dp)
    
    def shortestDistanceAfterQueries(self, n: int, queries: List[List[int]]) -> List[int]:
        dp = [n-1-i for i in range(n)]
        tree = [[] for _ in range(n)]
        
        for i in range(n-1):
            tree[i+1].append(i)
            
        m = len(queries)
        res = [0] * m
        
        for i, q in enumerate(queries):
            a, b = q[0], q[1]
            tree[b].append(a)
            dp[a] = min(dp[a], dp[b] + 1)
            self.dfs(tree, a, dp)
            res[i] = dp[0]
            
        return res

File: solutions/3517-shortest-distance-after-road-addition-queries-i/shortest-distance-after-road-addition-queries-i.java
class Solution {
    private void dfs(List<List<Integer>> tree, int c, int[] dp) {
        int d = dp[c] + 1;
        for (int x : tree.get(c)) {
            if (dp[x] <= d) continue;
            dp[x] = d;
            dfs(tree, x, dp);
        }
    }
    
    public int[] shortestDistanceAfterQueries(int n, int[][] queries) {
        int[] dp = new int[n];
        for (int i = 0; i < n; i++) {
            dp[i] = n - 1 - i;
        }
        
        List<List<Integer>> tree = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            tree.add(new ArrayList<>());
        }
        
        for (int i = 0; i < n - 1; i++) {
            tree.get(i + 1).add(i);
        }
        
        int m = queries.length;
        int[] res = new int[m];
        
        for (int i = 0; i < m; i++) {
            int a = queries[i][0];
            int b = queries[i][1];
            tree.get(b).add(a);
            dp[a] = Math.min(dp[a], dp[b] + 1);
            dfs(tree, a, dp);
            res[i] = dp[0];
        }
        
        return res;
    }
}

